### PR TITLE
Support preloading of native epoll lib

### LINKED ISSUE #5122: Allow Netty-epoll use from preloaded System library
In the same vein as #5043, it would be really useful if there was some way to use Netty's epoll without it doing the library initialization itself.   The fix would be pretty much the same as here: 379ad2c02ed0c0ae9f94e4081e3f910ece6380b7

cc: @nmittler

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

Some applications may use alternative methods of loading the epoll JNI symbols. We should support this use case.

Modifications:

Attempt to use a side effect free JNI method. If that fails, load the library.

Result:

Fixes #5122

### CODE REVIEW

**Comment by nmittler** [comment]
@carl-mastrangelo @normanmaurer @Scottmitch PTAL

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java [hunk 1]
```diff
@@ -53,14 +53,16 @@
...
 public final class Native {
     static {
-        String name = SystemPropertyUtil.get("os.name").toLowerCase(Locale.UK).trim();
-        if (!name.startsWith("linux")) {
-            throw new IllegalStateException("Only supported on Linux");
+        try {
+            // First, try calling a side-effect free JNI method to see if the library was already
+            // loaded by the application.
+            offsetofEpollData();
+        } catch (UnsatisfiedLinkError e) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java]: e -> ignore

**Comment by normanmaurer** [comment]
@nmittler only one nit...

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java (hunk 1 above)
Comment (nmittler) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java]: done.

**Comment by normanmaurer** [comment]
LGTM

**Comment by carl-mastrangelo** [comment]
@nmittler LGTM, Thanks for taking care of this!

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java [hunk 2]
```diff
@@ -250,4 +252,14 @@ public static void setTcpMd5Sig(int fd, InetAddress address, byte[] key) throws
...
         // utility
     }
+
+    private static void loadNativeLibrary() {
+        String name = SystemPropertyUtil.get("os.name").toLowerCase(Locale.UK).trim();
+        if (!name.startsWith("linux")) {
+            throw new IllegalStateException("Only supported on Linux");
+        }
+        NativeLibraryLoader.load(SystemPropertyUtil.get("io.netty.packagePrefix", "").replace('.', '-') +
+                "netty-transport-native-epoll",
```
Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java]: nit: fix indentation of these params

**Comment by Scottmitch** [comment]
Leak during build failure appears to be unrelated and was captured in https://github.com/netty/netty/issues/5127

**Comment by Scottmitch** [comment]
@nmittler - 1 nit and lgtm

**Comment by normanmaurer** [comment]
@Scottmitch the leak was fixed a few hours ago by me...

**Comment by normanmaurer** [comment]
@nmittler once you addressed @Scottmitch s comment, ship it! :)

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java (hunk 2 above)
Comment (nmittler) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java]: done

**Comment by nmittler** [comment]
cherry-picked into 4.1 (de2515d)

**Comment by normanmaurer** [comment]
Please also into 4.0

**Comment by normanmaurer** [comment]
Nevermind I did it for you as d9b4ec050caa6178e38e6c54eed37cb5ef5083d9 :)

**Pull request closed without merging**